syncserver/src/lib.rs

```diff
@@ -3,12 +3,243 @@
 
 #[macro_use]
 extern crate slog_scope;
-#[macro_use]
-extern crate validator_derive;
 
-#[macro_use]
-pub mod error;
 pub mod logging;
-pub mod server;
-pub mod tokenserver;
-pub mod web;
+mod middleware;
+
+use std::{error::Error, sync::Arc};
+
+use actix_cors::Cors;
+#[cfg(feature = "syncstorage")]
+use actix_web::http::StatusCode;
+#[cfg(feature = "syncstorage")]
+use actix_web::middleware::errhandlers::ErrorHandlers;
+use actix_web::{dev, http::Method, web, App, HttpServer};
+#[cfg(not(all(feature = "syncstorage", feature = "tokenserver")))]
+use actix_web::{web::ServiceConfig, HttpRequest, HttpResponse};
+use syncserver_settings::Settings;
+
+#[cfg(not(any(feature = "syncstorage", feature = "tokenserver")))]
+compile_error!("At least one of the \"syncstorage\" or \"tokenserver\" features must be enabled for this crate.");
+
+pub struct Server;
+
+#[cfg(all(feature = "syncstorage", not(feature = "tokenserver")))]
+#[macro_export]
+macro_rules! build_app {
+    ($settings: expr) => {
+        App::new()
+            .data(Arc::new($settings.master_secret.clone()))
+            // Middleware is applied LIFO
+            // These will wrap all outbound responses with matching status codes.
+            .wrap(ErrorHandlers::new().handler(
+                StatusCode::NOT_FOUND,
+                syncstorage_web::middleware::render_404,
+            ))
+            // These are our wrappers
+            .wrap(syncstorage_web::middleware::WeaveTimestamp::default())
+            .wrap(middleware::sentry::SentryWrapper::<
+                syncstorage_web::ServerState,
+                syncstorage_web::ApiError,
+            >::default())
+            .wrap(middleware::rejectua::RejectUA::<syncstorage_web::ServerState>::default())
+            // Followed by the "official middleware" so they run first.
+            // actix is getting increasingly tighter about CORS headers. Our server is
+            // not a huge risk but does deliver XHR JSON content.
+            // For now, let's be permissive and use NGINX (the wrapping server)
+            // for finer grained specification.
+            .wrap(build_cors($settings))
+            .wrap(syncstorage_web::middleware::EmitTokenserverOrigin::default())
+            .configure(
+                syncstorage_web::get_configurator(
```

Comment:
The idea here is that `syncstorage-web` and `tokenserver-web` will each have their own `get_configurator` function, which is responsible for registering the routes specific to each of the two services. The `syncserver` crate doesn't need to know anything about the particular implementations in order to make use of them; `syncserver`'s role is to decide whether to include Syncstorage, Tokenserver, or both, depending on what cargo features are set